packages/nimble-components/src/menu-button/index.ts

```diff
@@ -162,16 +170,61 @@ export class MenuButton extends FoundationElement implements ButtonPattern {
     public menuKeyDownHandler(e: KeyboardEvent): boolean {
         switch (e.key) {
             case keyEscape:
-                this.open = false;
+                this.setOpen(false);
                 this.toggleButton!.focus();
                 return false;
             default:
                 return true;
         }
     }
 
+    private setOpen(newValue: boolean): void {
+        if (this.open === newValue) {
+            return;
+        }
+
+        const eventDetail: MenuButtonToggleEventDetail = {
+            oldState: this.open,
+            newState: newValue
+        };
+        this.$emit('beforetoggle', eventDetail);
+
+        this.open = newValue;
+    }
+
     private get menu(): HTMLElement | undefined {
```

Comment:
nit / pref: this is doing a fair amount of work at this point so I'd probably say lets go ahead and make this a `getMenu()` method instead